File: app/src/test/java/com/test/xyz/demo/domain/interactor/project/ProjectInteractorTest.java
package com.test.xyz.demo.domain.interactor.project;

import com.test.xyz.demo.domain.model.github.GitHubRepo;
import com.test.xyz.demo.domain.repository.api.ProjectListRepository;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import io.reactivex.Observable;
import io.reactivex.plugins.RxJavaPlugins;
import io.reactivex.schedulers.Schedulers;

import static org.mockito.AdditionalMatchers.not;
import static org.mockito.Matchers.anyString;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.when;

public class ProjectInteractorTest {
    private static final String USER_NAME = "hazems";
    private static final String PROJECT_ID = "Test";
    private static final String UNLUCKY_ACCOUNT = "UNLUCKY_ACCOUNT";
    private GitHubRepo gitHubRepo;
    private List<GitHubRepo> gitHubRepoList;

    @Mock ProjectListRepository projectListRepository;

    ProjectInteractorImpl testSubject;

    @Before
    public void setup() {
        MockitoAnnotations.initMocks(this);
        initializeRxSchedulers();
        initializeFakeGitHubRepos();
        testSubject = new ProjectInteractorImpl(projectListRepository);
    }

    @Test
    public void getProjectList_whenUserNameIsCorrect_shouldReturnRepoListInfo() {
        //GIVEN
        mockGetProjectListAPI();

        //WHEN
        Observable<List<GitHubRepo>> projectListObservable = testSubject.getProjectList(USER_NAME);

        //THEN
        projectListObservable.test()
                .assertValues(gitHubRepoList)
                .assertNoErrors();
    }

    @Test
    public void getProjectList_whenUserNameIsEmpty_shouldReturnValidationError() {
        //GIVEN
        mockGetProjectListAPI();

        //WHEN
        Observable<List<GitHubRepo>> projectListObservable = testSubject.getProjectList("");

        //THEN
        projectListObservable.test()
                .assertError(IllegalArgumentException.class);
    }

    @Test
    public void getProjectList_whenNetworkErrorHappen_shouldReturnFailureError() {
        //GIVEN
        mockGetProjectListAPI();

        //WHEN
        Observable<List<GitHubRepo>> projectListObservable = testSubject.getProjectList(UNLUCKY_ACCOUNT);

        //THEN
        projectListObservable.test()
                .assertError(IOException.class);
    }

    @Test
    public void getProjectDetails_whenUserNameAndProjectIDAreCorrect_shouldReturnRepoItemInfo() {
        //GIVEN
        mockGetProjectDetailsAPI();

        //WHEN
        Observable<GitHubRepo> projectDetailsObservable = testSubject.getProjectDetails(USER_NAME, PROJECT_ID);

        //THEN
        projectDetailsObservable.test()
                .assertValues(gitHubRepo)
                .assertNoErrors();
    }

    @Test
    public void getProjectDetails_whenUserNameIsEmpty_shouldReturnValidationError() {
        //GIVEN
        mockGetProjectDetailsAPI();

        // WHEN
        Observable<GitHubRepo> projectDetailsObservable = testSubject.getProjectDetails("", PROJECT_ID);

        //THEN
        projectDetailsObservable.test()
                .assertError(IllegalArgumentException.class);
    }

    @Test
    public void getProjectDetails_whenProjectIDIsEmpty_shouldReturnValidationError() {
        //GIVEN
        mockGetProjectDetailsAPI();

        // WHEN
        Observable<GitHubRepo> projectDetailsObservable = testSubject.getProjectDetails("", PROJECT_ID);

        //THEN
        projectDetailsObservable.test()
                .assertError(IllegalArgumentException.class);
    }

    @Test
    public void getProjectDetails_whenNetworkErrorHappen_shouldReturnFailureError() {
        //GIVEN
        mockGetProjectDetailsAPI();

        // WHEN
        Observable<GitHubRepo> projectDetailsObservable = testSubject.getProjectDetails(UNLUCKY_ACCOUNT, PROJECT_ID);

        //THEN
        projectDetailsObservable.test()
                .assertError(IOException.class);
    }

    private void mockGetProjectListAPI() {
        mockGetProjectListHappyPath();
        mockGetProjectListErrorPath();
    }

    private void mockGetProjectDetailsAPI() {
        mockGetProjectDetailsHappyPath();
        mockGetProjectDetailsErrorPath();
    }

    private void mockGetProjectListHappyPath() {
        Observable<List<GitHubRepo>> observable = Observable.just(gitHubRepoList);
        when(projectListRepository.getProjectList(not(eq(UNLUCKY_ACCOUNT)))).thenReturn(observable);
    }

    private void mockGetProjectListErrorPath() {
        when(projectListRepository.getProjectList(eq(UNLUCKY_ACCOUNT)))
                .thenReturn(Observable.error(new IOException("Invalid account")).cast((Class) List.class));
    }

    private void mockGetProjectDetailsHappyPath() {
        Observable<GitHubRepo> observable = Observable.just(gitHubRepo);
        when(projectListRepository.getProjectDetails(not(eq(UNLUCKY_ACCOUNT)), anyString())).thenReturn(observable);
    }

    private void mockGetProjectDetailsErrorPath() {
        when(projectListRepository.getProjectDetails(eq(UNLUCKY_ACCOUNT), anyString()))
                .thenReturn(Observable.error(new IOException("Invalid account")).cast((Class) List.class));
    }

    private void initializeFakeGitHubRepos() {
        gitHubRepo = new GitHubRepo("RepoItem");
        gitHubRepoList = new ArrayList<>();

        for (int i = 0; i < 10; ++i) {
            gitHubRepoList.add(new GitHubRepo("SampleRepoItem" + i));
        }
    }

    private void initializeRxSchedulers() {
        RxJavaPlugins.setIoSchedulerHandler(scheduler -> Schedulers.trampoline());
    }
}

File: app/src/test/kotlin/com/test/xyz/demo/domain/interactor/project/ProjectInteractorTest.kt
package com.test.xyz.demo.domain.interactor.project

import com.nhaarman.mockitokotlin2.doReturn
import com.nhaarman.mockitokotlin2.eq
import com.nhaarman.mockitokotlin2.mock
import com.test.xyz.demo.domain.model.github.GitHubRepo
import com.test.xyz.demo.domain.repository.api.ProjectListRepository
import io.reactivex.Observable
import io.reactivex.plugins.RxJavaPlugins
import io.reactivex.schedulers.Schedulers
import org.junit.Before
import org.junit.Test
import org.mockito.ArgumentMatchers.anyString
import org.mockito.MockitoAnnotations
import java.io.IOException

class ProjectInteractorTest {
    val gitHubRepo = GitHubRepo(1, "RepoItem")
    val gitHubRepoList = arrayListOf(GitHubRepo(0, "Fake GitHub Repo-0"),
            GitHubRepo(1, "Fake GitHub Repo-1"),
            GitHubRepo(2, "Fake GitHub Repo-2"))

    val projectListRepository = mock<ProjectListRepository> {
        on {getProjectList(USER_NAME)}                           doReturn Observable.just<List<GitHubRepo>>(gitHubRepoList)
        on {getProjectList(eq(INVALID_ACCOUNT))}                 doReturn Observable.error<List<GitHubRepo>>(IOException("Invalid account"))
        on {getProjectDetails(eq(USER_NAME), anyString())}       doReturn Observable.just(gitHubRepo)
        on {getProjectDetails(eq(INVALID_ACCOUNT), anyString())} doReturn Observable.error<GitHubRepo>(IOException("Invalid account"))
    }

    private lateinit var testSubject: ProjectInteractorImpl

    @Before
    fun setup() {
        init()
        testSubject = ProjectInteractorImpl(projectListRepository)
    }

    @Test
    fun `getProjectList() should return repo list information`() {
        //WHEN
        val projectListObservable = testSubject.getProjectList(USER_NAME)

        //THEN
        projectListObservable.test()
                .assertValues(gitHubRepoList)
                .assertNoErrors()
    }

    @Test
    fun `getProjectList(), when user name is empty, it should return a validation error`() {
        //WHEN
        val projectListObservable = testSubject.getProjectList("")

        //THEN
        projectListObservable.test()
                .assertError(IllegalArgumentException::class.java)
    }

    @Test
    fun `getProjectList(), when network error happens, it should return a failure error`() {
        //WHEN
        val projectListObservable = testSubject.getProjectList(INVALID_ACCOUNT)

        //THEN
        projectListObservable.test()
                .assertError(IOException::class.java)
    }

    @Test
    fun `getProjectDetails() should return a Repo item Info`() {
        //WHEN
        val projectDetailsObservable = testSubject.getProjectDetails(USER_NAME, PROJECT_ID)

        //THEN
        projectDetailsObservable.test()
                .assertValues(gitHubRepo)
                .assertNoErrors()
    }

    @Test
    fun `getProjectDetails(), when user name is empty, it should return a validation error`() {
        // WHEN
        val projectDetailsObservable = testSubject.getProjectDetails("", PROJECT_ID)

        //THEN
        projectDetailsObservable.test()
                .assertError(IllegalArgumentException::class.java)
    }

    @Test
    fun `getProjectDetails(), when project is is empty, it should return a validation error`() {
        // WHEN
        val projectDetailsObservable = testSubject.getProjectDetails("", PROJECT_ID)

        //THEN
        projectDetailsObservable.test()
                .assertError(IllegalArgumentException::class.java)
    }

    @Test
    fun `getProjectDetails(), when a network error happens, it should return a failure error`() {
        // WHEN
        val projectDetailsObservable = testSubject.getProjectDetails(INVALID_ACCOUNT, PROJECT_ID)

        //THEN
        projectDetailsObservable.test()
                .assertError(IOException::class.java)
    }

    private fun init() {
        MockitoAnnotations.initMocks(this)
        RxJavaPlugins.setIoSchedulerHandler { scheduler -> Schedulers.trampoline() }
    }

    companion object {
        private val USER_NAME = "hazems"
        private val PROJECT_ID = "Test"
        private val INVALID_ACCOUNT = "INVALID_ACCOUNT"
    }
}
